src/main/host/thread_ptrace.c

```diff
@@ -1024,7 +1024,15 @@ SysCallCondition* threadptrace_resume(Thread* base) {
             process_flushPtrs(thread->base.process);
 
             trace("ptrace resuming with signal %ld", thread->signalToDeliver);
-            // Allow child to start executing.
+            // Allow child to start executing.  We release the shared memory
+            // lock, which we will reacquire when the child stops again.
+            //
+            // XXX: This is currently a bit fragile - stopping and starting the
+            // managed thread ideally should be refactored into helper functions
+            // that also take and release the lock. Not worth it for now since
+            // we plan to remove ptrace mode
+            // https://github.com/shadow/shadow/issues/1945.
+            host_unlockShimShmemLock(thread->base.host);
             if (ptrace(PTRACE_SYSEMU, thread->base.nativeTid, 0, thread->signalToDeliver) < 0) {
                 utility_panic("ptrace %d: %s", thread->base.nativeTid, g_strerror(errno));
                 return NULL;
```

Comment:
Add a `host_lockShimShmemLock()` before the `return NULL` for non-debug builds?

Reply:
Oh wow, I'd forgotten that `utility_panic` is a no-op in non-debug builds. For now just changed this `return` to an `abort`.